rfd/0112-automatic-database-users.md

```diff
@@ -0,0 +1,317 @@
+updated with `nologin` trait so they can't connect.
+
+### Stored procedures
+
+To create and disable users Teleport will use 2 stored procedures:
+
+- `teleport_create_user(username varchar, roles varchar[])`
+- `teleport_disable_user(username varchar)`
+
+Usage of stored procedures is chosen for the following reasons:
+
+- Utility statements like create/grant do not support bound parameters so
+  using procedure is safer that building queries in the code. See security
+  section below for more details.
+- Stored procedures are "all or nothing" so if any of the roles are invalid
+  during user creation for example, the user won't be created and no extra
+  cleanup will be required.
+- Stored procedures make it easier to keep some of the control logic in the
+  database for avoid races between agents, for example only disabling the user
+  if it doesn't have any active sessions.
+
+The procedures' source code is included below.
+
+### Session flow
+
+When receiving a database client connection, Teleport database service will
+first connect to the database as the "adminUser", install both procedures
+described above and call `teleport_create_user()` to provision a user and
+assign it roles, or reactivate the user that was previously disabled by
+Teleport.
+
+After that, Teleport will proceed to connecting to the database as usual using
+the newly created user.
+
+If such user already exists in the database, Teleport will connect as the
+existing user (without executing grant statements). This will allow multiple
+users to connect as the same auto-created user, or to connect as an already
+existing user.
+
+After the database session completes, database service will connect again as
+the "adminUser" and call `teleport_disable_user()` to disable the created user
+unless it has other active sessions.
+
+### Create user procedure
+
+```sql
+create or replace procedure teleport_create_user(username varchar, roles varchar[])
+language plpgsql
+as $$
+declare
+    role_ varchar;
+begin
+    -- If the user already exists and was provisioned by Teleport, reactivate it,
+    -- otherwise provision it.
+    if exists (select * from pg_auth_members where
+                roleid = (select oid from pg_roles where rolname = 'teleport-auto-user') and
+                member = (select oid from pg_roles where rolname = username)
+    ) then
+        -- Utility statements like create/grant do not accept positional arguments
+        -- so we have to use dynamic SQL. %I takes care of safe argument escaping.
+        -- For RDS we will automatically add "rds_iam" to the roles.
+        execute format('alter user %I with login', username);
+    else
+        execute format('create user %I in role teleport-auto-user', username);
+    end if;
+    -- Assign all roles to the created/activated user.
+    foreach role_ in array roles
+    loop
+        execute format('grant %I to %I', role_, username);
```

Comment:
Very good question, I think we should. Problem is, it would likely require some complicated logic to determine if the existing database account has same roles as you have or not, which I'm not too fond of putting in the stored procedure code 😄

I've got an idea: what if we don't actually use `db_users` for auto-provisioning and just use your Teleport username? So if your Teleport username is `alice@example.com`, you will get the same user inside database with your `db_roles`. Now that I think about it, there's really no reason to have multiple `db_users` if you're using auto-provisioning.

So every Teleport user will get their own database user with the same name which will solve the conflict problem, and the role configuration would be simpler as there wouldn't be a need to configure `db_users` for auto-provisioning. So I think it would simplify a lot of things and it will be easier to correlate Teleport users to database users as well.

@smallinsky @klizhentas What do you guys think?

Reply:
> I've got an idea: what if we don't actually use db_users for auto-provisioning and just use your Teleport username? So if your Teleport username is alice@example.com, you will get the same user inside database with your db_roles. Now that I think about it, there's really no reason to have multiple db_users if you're using auto-provisioning.

👍 I think that this is a good idea that will allow to simplify this solution. 
The only thing that we will need to consider that comes to my mind is fulfilling Postgres [username  field restrictions ](https://www.postgresql.org/docs/7.0/syntax525.htm#:~:text=Names%20in%20SQL%20must%20begin,but%20they%20will%20be%20truncated.)

- Names in SQL must begin with a letter (a-z) or underscore (_). Subsequent characters in a name can be letters, digits (0-9), or underscores. 
- The system uses no more than NAMEDATALEN-1 characters of a name; longer names can be written in queries, but they will be truncated.
-  By default, NAMEDATALEN is 32 so the maximum name length is 31

So we to handle teleport user identified by emails like: ` alice@example.com` we will need to introduce encoding or hashing. Additionally we should add some kind of validation for  `len(username) < NAMEDATALEN`

Reply:
@smallinsky Right, we might need to sanitize usernames a little bit and get rid of unsupported characters but Postgres actually does accept characters like `@` and `.` in the role names so I think we should be good. I tested it:

```
test=# create user "alice@example.com" with password '...';
CREATE ROLE
test=# \du
                                            List of roles
         Role name         |                         Attributes                         | Member of
---------------------------+------------------------------------------------------------+------------
 alice.bob                 |                                                            | {}
 alice@example.com         |                                                            | {}

➜  ~ psql postgres://alice%40example.com@localhost:5432/test
Password for user alice@example.com:
psql (14.2, server 13.5 (Debian 13.5-1.pgdg110+1))
Type "help" for help.
test=> \c
```